zokrates_core/src/static_analysis/uint_optimizer.rs

```diff
@@ -456,77 +456,70 @@ impl<'ast, T: Field> Folder<'ast, T> for UintOptimizer<'ast, T> {
-                    assert_eq!(lhs.len(), 1);
-                    self.register(
-                        lhs[0].clone(),
-                        UMetadata {
-                            max: T::from(2).pow(8) - T::from(1),
-                            should_reduce: ShouldReduce::False,
-                        },
-                    );
-                    vec![ZirStatement::MultipleDefinition(
-                        lhs,
-                        ZirExpressionList::EmbedCall(embed, generics, arguments),
-                    )]
-                }
-                _ => vec![ZirStatement::MultipleDefinition(
+            ) => {
+                match embed {
+                    FlatEmbed::U64FromBits => {
+                        assert_eq!(lhs.len(), 1);
+                        self.register(
+                            lhs[0].clone(),
+                            UMetadata {
+                                max: T::from(2).pow(64) - T::from(1),
+                                should_reduce: ShouldReduce::False,
+                            },
+                        );
+                    }
+                    FlatEmbed::U32FromBits => {
+                        assert_eq!(lhs.len(), 1);
+                        self.register(
+                            lhs[0].clone(),
+                            UMetadata {
+                                max: T::from(2).pow(32) - T::from(1),
+                                should_reduce: ShouldReduce::False,
+                            },
+                        );
+                    }
+                    FlatEmbed::U16FromBits => {
+                        assert_eq!(lhs.len(), 1);
+                        self.register(
+                            lhs[0].clone(),
+                            UMetadata {
+                                max: T::from(2).pow(16) - T::from(1),
+                                should_reduce: ShouldReduce::False,
+                            },
+                        );
+                    }
+                    FlatEmbed::U8FromBits => {
+                        assert_eq!(lhs.len(), 1);
+                        self.register(
+                            lhs[0].clone(),
+                            UMetadata {
+                                max: T::from(2).pow(8) - T::from(1),
+                                should_reduce: ShouldReduce::False,
+                            },
+                        );
+                    }
+                    _ => {}
+                };
+
+                vec![ZirStatement::MultipleDefinition(
                     lhs,
                     ZirExpressionList::EmbedCall(
                         embed,
                         generics,
                         arguments
                             .into_iter()
-                            .map(|e| self.fold_expression(e))
+                            .map(|e| match e {
+                                ZirExpression::Uint(e) => {
```

Comment:
@Schaeff this needs a closer look

Reply:
Let's always reduce

Reply:
Then in `flatten_u_to_bits` we can assert that `from == to` in terms of bitwidths and just call `get_bits`